packages/tailwindcss/src/index.ts

```diff
@@ -113,7 +113,11 @@ export function compile(
       customUtilities.push((designSystem) => {
         designSystem.utilities.static(name, (candidate) => {
           if (candidate.negative) return
-          return structuredClone(node.nodes)
+          let ast = structuredClone(node.nodes)
+
+          substituteAtApply(ast, designSystem)
```

Comment:
Feels a bit weird to me to make this the responsibility of the actual plugin/matcher/whatever 🤔

Should we maybe just handle all of the `@apply` stuff in a separate pass at the end? Or perhaps we could replace it in the CSS up front, so we take this:

```css
@utility foo {
  @apply flex
}
```

...and turn that into:

```css
@utility foo {
  & {
    display: flex
  }
}
```

...before the plugin is even registered with the design system?

Reply:
I think this would cause problems when a custom utility is used in an `@apply` inside another custom utility.

Reply:
Doing this as a separate pass seems reasonable. We could probably:
- Perform a single pass in `compile()` — making sure there's less to replace later
- Perform a pass every time we call `build()` — and we'd only ever need to do this for the _new_ nodes.

Reply:
I 100% agree, but these are also lazily evaluated, so it will only take effect the moment we actually use the utility. If you never use them in the first place, handling the `@apply` is a bit wasteful.

> Replace `@apply` in a separate pass at the end

That's an option, but as long as the CSS files don't change, the `@apply` directives used in the CSS itself won't change either. So doing this as a separate pass whenever we build the final CSS based on the candidates seems a bit wasteful. (Because then we would do it `n` times for every `build()` instead of once)

The `@apply` used in a `@utility` should only be handled if the utility is being used somewhere, so this is only available once we do the `build(candidates)` call.

> Or perhaps we could replace it in the CSS up front

We can do that for _some_ utilities (the built-in ones, and custom `@utility` ones that were registered before we see the first `@apply`). If you have this setup:

```css
@utility foo {
  @apply hover:bar;
}

@utility bar {
  color: red;
}
```

Then we can't replace the `@apply hover:bar` in `foo` yet because the `bar` utility has not been registered yet.

Another issue is that we won't be able to detect circular references this way. If `bar` happens to reference `foo` then we should detect that.

Reply:
We can however collect the nodes and once all custom `@utility` utilities are registered, replace the `@apply` once and later we can re-use the replaced `@apply` nodes.

Reply:
I think I figured out a nice way of handling this (https://github.com/tailwindlabs/tailwindcss/pull/14144/commits/48ff0b8884e23b546f377fc4c10ee312166bffb6)

1. Collect the `@utility` rules _if_ they contain `@apply`
2. Register the `@utility` in the design sytem
3. Once all `@utility` utilities are registered, substitute the `@apply` with its correct rule. This way we don't have to handle the `@apply` over and over and over again, just once.
4. Bonus points, since we register all utilities first, we can apply `@utility` out of order and detect circular dependencies.